app/src/main/java/com/example/habithelper/activities/LoginActivity.java

```diff
@@ -47,23 +45,21 @@ public void onClick(View v) {
             }
         });
 
-        btnLogin.setOnClickListener(new View.OnClickListener(){
+        btnLogin.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View v) {
                 String username = etUsername.getText().toString();
                 String password = etPassword.getText().toString();
                 loginUser(username, password);
             }
         });
-
-
     }
 
-    private void loginUser(String username, String password){
+    private void loginUser(String username, String password) {
         ParseUser.logInInBackground(username, password, new LogInCallback() {
             @Override
             public void done(ParseUser user, ParseException e) {
```

Comment:
Not familiar with parser, and certainly agree that ideally there should be onSuccess and onError flows to avoid having exceptions on success, but I recall seeing this on the codepath material. I would recommend to check documentation if it can be implemented as the suggested onSuccess and onError as a stretch goal. Maybe file an issue to avoid forgetting it?

Reply:
Good idea! This is a Parse built in function.

Reply:
It automatically throws ParseExceptionErrors, but this will go on my list of stretch features for sure.